File: Freedom/微信/基本类库/Common/WXPictureCarouselView.m
//  TLPictureCarouselView.m
//  Freedom
// Created by Super
#import "WXPictureCarouselView.h"
#import <BlocksKit/BlocksKit.h>
@interface WXPictureCarouselViewCell : UICollectionViewCell
@property (nonatomic, strong) id<WXPictureCarouselProtocol> model;
@end
@interface WXPictureCarouselViewCell ()
@property (nonatomic, strong) UIImageView *imageView;
@end
@implementation WXPictureCarouselViewCell
- (id)initWithFrame:(CGRect)frame{
    if (self = [super initWithFrame:frame]) {
        [self.contentView addSubview:self.imageView];
        
        [self p_addMasonry];
    }
    return self;
}
- (void)setModel:(id<WXPictureCarouselProtocol>)model{
    [self.imageView sd_setImageWithURL:[NSURL URLWithString:[model pictureURL]]];
}
#pragma mark - 
- (void)p_addMasonry{
    [self.imageView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(self.contentView);
    }];
}
#pragma mark - 
- (UIImageView *)imageView{
    if (_imageView == nil) {
        _imageView = [[UIImageView alloc] init];
    }
    return _imageView;
}
@end
@interface WXPictureCarouselView () <UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout>
@property (nonatomic, strong) NSTimer *timer;
@property (nonatomic, strong) UICollectionView *collectionView;
@end
@implementation WXPictureCarouselView
- (id)initWithFrame:(CGRect)frame{
    if (self = [super initWithFrame:frame]) {
        self.timeInterval = DEFAULT_TIMEINTERVAL;
        [self addSubview:self.collectionView];
        
        [self p_addMasonry];
        
        [self.collectionView registerClass:[WXPictureCarouselViewCell class] forCellWithReuseIdentifier:@"TLPictureCarouselViewCell"];
    }
    return self;
}
- (void)setData:(NSArray *)data{
    _data = data;
    [self.collectionView reloadData];
    dispatch_async(dispatch_get_main_queue(), ^{
        
        CGPoint offset = CGPointMake(self.collectionView.frame.size.width * 1,self.collectionView.contentOffset.y);
        [self.collectionView setContentOffset:offset animated:NO];
        
        if (self.timer == nil && self.data.count > 1) {
            __weak typeof(self) weakSelf = self;
            self.timer = [NSTimer bk_scheduledTimerWithTimeInterval:2.0 block:^(NSTimer *tm) {
                [weakSelf scrollToNextPage];
            } repeats:YES];
        }
    });
}
- (void)dealloc{
    [self.timer invalidate];
    self.timer = nil;
}
#pragma mark - 
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    return self.data.count == 0 ? 0 : self.data.count + 2;
}
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    NSInteger row = indexPath.row == 0 ? self.data.count - 1 : (indexPath.row == self.data.count + 1 ? 0 : indexPath.row - 1);
    id<WXPictureCarouselProtocol> model = self.data[row];
    WXPictureCarouselViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"TLPictureCarouselViewCell" forIndexPath:indexPath];
    [cell setModel:model];
    return cell;
}
- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath{
    NSInteger row = indexPath.row == 0 ? self.data.count - 1 : (indexPath.row == self.data.count - 1 ? 0 : indexPath.row - 1);
    id<WXPictureCarouselProtocol> model = self.data[row];
    if (self.delegate && [self.delegate respondsToSelector:@selector(pictureCarouselView:didSelectItem:)]) {
        [self.delegate pictureCarouselView:self didSelectItem:model];
    }
}
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath{
    return collectionView.frame.size;
}
//MARK: UIScrollViewDelegate
- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView{
    [self.timer invalidate];
    self.timer = nil;
}
- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate{
    if (self.timer == nil && self.data.count > 1) {
        __weak typeof(self) weakSelf = self;
        self.timer = [NSTimer bk_scheduledTimerWithTimeInterval:2.0 block:^(NSTimer *tm) {
            [weakSelf scrollToNextPage];
        } repeats:YES];
    }
}
- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView{
    // 轮播实现
    
    if (scrollView.contentOffset.x / scrollView.frame.size.width == 0) {
        CGPoint offset = CGPointMake(self.collectionView.frame.size.width * self.data.count,self.collectionView.contentOffset.y);
        [scrollView setContentOffset:offset animated:NO];
    }else if (scrollView.contentOffset.x / scrollView.frame.size.width == self.data.count + 1) {
        CGPoint offset = CGPointMake(self.collectionView.frame.size.width * 1,self.collectionView.contentOffset.y);
        [scrollView setContentOffset:offset animated:NO];
    }
}
#pragma mark - Event Response
- (void)scrollToNextPage{
    NSInteger nextPage;
    if (self.collectionView.contentOffset.x / self.collectionView.frame.size.width == self.data.count) {
        CGPoint offset = CGPointMake(self.collectionView.frame.size.width * 0,self.collectionView.contentOffset.y);
        [self.collectionView setContentOffset:offset animated:NO];
        nextPage = 1;
    }else {
        nextPage = self.collectionView.contentOffset.x / self.collectionView.frame.size.width + 1;
    }
    CGPoint offset = CGPointMake(self.collectionView.frame.size.width * nextPage,self.collectionView.contentOffset.y);
    [self.collectionView setContentOffset:offset animated:YES];
}
#pragma mark - 
- (void)p_addMasonry{
    [self.collectionView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(self);
    }];
}
#pragma mark - 
- (UICollectionView *)collectionView{
    if (_collectionView == nil) {
        UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc] init];
        [layout setScrollDirection:UICollectionViewScrollDirectionHorizontal];
        [layout setMinimumLineSpacing:0];
        [layout setMinimumInteritemSpacing:0];
        [layout setSectionInset:UIEdgeInsetsZero];
        _collectionView = [[UICollectionView alloc] initWithFrame:CGRectZero collectionViewLayout:layout];
        [_collectionView setBackgroundColor:[UIColor whiteColor]];
        [_collectionView setShowsVerticalScrollIndicator:NO];
        [_collectionView setShowsHorizontalScrollIndicator:NO];
        [_collectionView setPagingEnabled:YES];
        [_collectionView setScrollsToTop:NO];
        [_collectionView setDelegate:self];
        [_collectionView setDataSource:self];
    }
    return _collectionView;
}
@end

File: Freedom/微信/基本类库/Common/WXPictureCarouselView.swift
//
//  WXPictureCarouselView.swift
//  Freedom
import SnapKit
import Foundation
protocol WXPictureCarouselDelegate: NSObjectProtocol {
    func pictureCarouselView(_ pictureCarouselView: WXPictureCarouselView, didSelectItem model: WXPictureCarouselProtocol)
}
class WXPictureCarouselViewCell: UICollectionViewCell {
    var model: WXPictureCarouselProtocol? {
        didSet {
            if let url = URL(string: model?.pictureURL() ?? "") {
                imageView.sd_setImage(with: url)
            }
        }
    }
    private var imageView = UIImageView()
    override init(frame: CGRect) {
        super.init(frame: frame)
        contentView.addSubview(imageView)
        imageView.snp.makeConstraints { (make) in
            make.edges.equalTo(self.contentView)
        }
    }
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
}
class WXPictureCarouselView:UIView, UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {
    private var timer: Timer?
    private lazy var collectionView: UICollectionView = {
        let layout = UICollectionViewFlowLayout()
        layout.scrollDirection = .horizontal
        layout.minimumLineSpacing = 0
        layout.minimumInteritemSpacing = 0
        layout.sectionInset = .zero
        let collectionView = UICollectionView(frame: CGRect.zero, collectionViewLayout: layout)
        collectionView.backgroundColor = UIColor.white
        collectionView.showsVerticalScrollIndicator = false
        collectionView.showsHorizontalScrollIndicator = false
        collectionView.isPagingEnabled = true
        collectionView.scrollsToTop = false
        collectionView.delegate = self
        collectionView.dataSource = self
        return collectionView
    }()
    var data: [WXPictureCarouselProtocol] = [] {
        didSet {
            collectionView.reloadData()
            DispatchQueue.main.async(execute: {
                let offset = CGPoint(x: self.collectionView.frame.size.width * 1, y: self.collectionView.contentOffset.y)
                self.collectionView.setContentOffset(offset, animated: false)

                if self.timer == nil && self.data.count > 1 {
                    self.timer = Timer(fire: Date(), interval: 2, repeats: true, block: { (tm) in
                        self.scrollToNextPage()
                    })
                }
            })
        }
    }
    weak var delegate: WXPictureCarouselDelegate?
    var timeInterval: TimeInterval = 5
    override init(frame: CGRect) {
        super.init(frame: frame)
        addSubview(collectionView)
        collectionView.register(WXPictureCarouselViewCell.self, forCellWithReuseIdentifier: WXPictureCarouselViewCell.identifier)
        collectionView.snp.makeConstraints { (make) in
            make.edges.equalToSuperview()
        }
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    deinit {
        timer?.invalidate()
        timer = nil
    }
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return data.count == 0 ? 0 : data.count + 2
    }
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let row: Int = indexPath.row == 0 ? data.count - 1 : (indexPath.row == data.count + 1 ? 0 : indexPath.row - 1)
        let model: WXPictureCarouselProtocol = data[row]
        let cell = collectionView.dequeueCell(WXPictureCarouselViewCell.self, for: indexPath)
        cell.model = model
        return cell
    }
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let row: Int = indexPath.row == 0 ? data.count - 1 : (indexPath.row == data.count - 1 ? 0 : indexPath.row - 1)
        let model: WXPictureCarouselProtocol = data[row]
        delegate?.pictureCarouselView(self, didSelectItem: model)
    }

    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return collectionView.frame.size
    }
    func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
        timer?.invalidate()
        timer = nil
    }
    func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        if timer == nil && data.count > 1 {
            timer = Timer(fire: Date(), interval: 2.0, repeats: true, block: { (tm) in
                self.scrollToNextPage()
            })
        }
    }
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        if scrollView.contentOffset.x / scrollView.frame.size.width == 0 {
            let offset = CGPoint(x: collectionView.frame.size.width * CGFloat(data.count), y: collectionView.contentOffset.y)
            scrollView.setContentOffset(offset, animated: false)
        } else if Int(scrollView.contentOffset.x / scrollView.frame.size.width) == data.count + 1 {
            let offset = CGPoint(x: collectionView.frame.size.width * 1, y: collectionView.contentOffset.y)
            scrollView.setContentOffset(offset, animated: false)
        }
    }
    func scrollToNextPage() {
        var nextPage: Int
        if Int(collectionView.contentOffset.x / collectionView.frame.size.width) == data.count {
            let offset = CGPoint(x: collectionView.frame.size.width * 0, y: collectionView.contentOffset.y)
            collectionView.setContentOffset(offset, animated: false)
            nextPage = 1
        } else {
            nextPage = Int(collectionView.contentOffset.x / collectionView.frame.size.width + 1)
        }
        let offset = CGPoint(x: collectionView.frame.size.width * CGFloat(nextPage), y: collectionView.contentOffset.y)
        collectionView.setContentOffset(offset, animated: true)
    }
}
